### PR TITLE
make dragging links do more

### PR DESCRIPTION
**Pull request by BurningTreeC**
this is my PR as result of #3148

### CODE REVIEW

**Comment by alexhough** [comment]
It would be good if dragging titles when they are displayed as links is also included in this.

As it stands when you drag a title with spaces into a new tiddler it is not wrapped in double square brackets like when you drag from the Open tab in sidebar.

There are two use cases: 

1. Drag a title onto an open tag pill to tag the dragged tiddler
2. Drag the title into the body of a tiddler above to create a link. In this case the double bracked option is required

**Comment by BurningTreeC** [comment]
Hi, with this new drag and drop you can drag tiddler titles, also if titles are shown as links, and they result in [[tiddler title]] if you drag from the open tab or from the tiddler title itself

How do you accomplish that tiddlers get tagged by dragging the titles onto tag pills? Is that behaviour default or is a change in the tagtemplate needed to make it work?

**Comment by BurningTreeC** [comment]
Oh, I just discovered a feature! I didn't know that dragging titles into the dropdown of tagpills taggs the tiddlers with that tag. I read that once in the documentation and didn't understand it correctly. This is nice but documentation isn't good

**Comment by alexhough** [comment]
@BurningTreeC  could the behviour be modified such that dragging a title into a new tiddler can transcluded?

in other words, rather that `[[Tiddler Title]]` we get `{{Tiddler Title}}`

**Comment by alexhough** [comment]
That's great! really useful

With Chrome the crl option results in a Chrome popup, but once you close that the drag n drop works as intended...

**Comment by BurningTreeC** [comment]
@alexhough 

> With Chrome the crl option results in a Chrome popup, but once you close that the drag n drop works as intended...

Could you describe what the popup was displaying? That'd be very helpful, I'm not getting popus or alerts in firefox, chrome, chromium or TiddlyDesktop

> Could you hold down another key, drag the link then create a new tiddler tagged with that tiddler.

Sure we could, the possibilities would be many. BUT

we've decided to go with the three, because linking, transclusion and prefix - suffix are the most essential parts of the wikitext syntax

What you describe, the dragging of links somewhere to tag, is already possible using the droppable widget, no need to add something to this pull request

>  drag a missing link from the preview pane while holding down a key to open that tiddler in the story in edit mode...

Well, clicking it would do the same, doesn't it?

all the best, Simon

**Comment by Jermolene** [comment]
Hi @BurningTreeC do you think it makes sense to wait for any more feedback on your [post on the dev mailing list](https://groups.google.com/d/msg/tiddlywiki/XsWMo7xrIzo/Tsh5-n0aBwAJ) before merge?

**Comment by BurningTreeC** [comment]
Hi @Jermolene ,
no I don't think so, it'd be ready to merge, but I'd like to change the keys-configuration because it works but using a single field for each modifier would be better. I have some time right now ... I'll give it a shot

**Comment by Jermolene** [comment]
Thanks @BurningTreeC

**Comment by BurningTreeC** [comment]
done, @Jermolene 
happy merging!

**Comment by Jermolene** [comment]
Thanks @BurningTreeC apologies for the further comments!

**Comment by BurningTreeC** [comment]
@Jermolene , no problem!

**Comment by BurningTreeC** [comment]
@Jermolene - sorry for asking, but am I supposed to see some comments somewhere?

Because there aren't any and you've mentioned further comments

**Comment by Jermolene** [comment]
Hi @BurningTreeC aaaaargh! Sorry, that's my fault; GitHub now requires a "submit” button to be clicked after making review comments, and I keep forgetting to click it....

File: core/modules/utils/dom/dragndrop.js [hunk 1]
```diff
@@ -181,4 +188,56 @@ function parseJSONTiddlers(json,fallbackTitle) {
 	return data;
 };
 
+function getTitleStringModified(dragAction,titleString,dragSettings) {
+	var hasBrackets = /^\[\[.*/.test(titleString);
+	switch(dragAction) {
+		case "plain":
+			titleString = hasBrackets ? titleString.replace(/\[/g, '').replace(/\]/g, '') : titleString ;
```
[the remark this answers is not in the document]
Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: you're right, it isn't safe. changing it to the following should cover this...

```
hasBrackets =  /^\[\[.*\]\]$/.test(titleString);
titleString.replace(/^\[\[/, '').replace(/\]\]$/, '')

```

>What are the situations where the titleString will already be bracketed?

Now I checked this, it's a relict of the beginning when I passed the stringified titleString. Now that I'm passing the titles array there's no need for it so it seems

I'll change that locally and try it

File: core/modules/utils/dom/dragndrop.js [hunk 2]
```diff
@@ -75,17 +76,23 @@ exports.makeDraggable = function(options) {
...
+					dataTransfer.clearData();
 				}
-				var jsonData = [];
+				var jsonData = [],
+					plainTiddlerTitles = $tw.wiki.filterTiddlers(plainTitles);
+				// Add corresponding plain titles to array after each data entry
 				if(titles.length > 1) {
-					titles.forEach(function(title) {
+					titles.forEach(function(title,index) {
+						var plainJsonTitle = JSON.stringify({ "\p\l\a\i\n" : plainTiddlerTitles[index] });
```
[the remark this answers is not in the document]
Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: This is my workaround for something I didn't find another solution for

The problem is that dragging something into draggable lists like the Open Tab, or just reordering such lists, or dragging a tag-pill with many tiddler titles attached into such a list, can cause bad results like 30 non-existent tiddlers open in the story river

The plain title must somehow be passed on for the droppable and dropzone widgets
Stuff can also be dragged from one wiki to the other, so I thought that the plain title must be added to the jsonData

I was hoping to get some comments or advice on this, so I made it a bit prominent with that \p\l\a\in
The droppable and dropzone widgets then check if this \p\a\i\n entry exists and use that as title

I would love to see a better solution but I'm not having it

File: core/modules/utils/dom/dragndrop.js [hunk 3]
```diff
@@ -181,4 +188,56 @@ function parseJSONTiddlers(json,fallbackTitle) {
...
 };
 
+function getTitleStringModified(dragAction,titleString,dragSettings) {
+	var hasBrackets = /^\[\[.*/.test(titleString);
+	switch(dragAction) {
+		case "plain":
+			titleString = hasBrackets ? titleString.replace(/\[/g, '').replace(/\]/g, '') : titleString ;
+			break;
+		case "transclude":
+			titleString = hasBrackets ? titleString.replace(/\[/g, '{').replace(/\]/g, '}') : '{{' + titleString + '}}' ;
```
[the remark this answers is not in the document]
Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: I'll check this too, see my comment to the first comment

File: core/modules/utils/dom/dragndrop.js [hunk 4]
```diff
@@ -181,4 +188,56 @@ function parseJSONTiddlers(json,fallbackTitle) {
...
+			titleString = !hasBrackets ? '[[' + titleString + ']]' : titleString ;
+	}
+	return titleString;
+};
+
+function dragModifiers(event,titleString) {
+	var dragSettings = $tw.wiki.getTiddler("$:/config/DragDefaults") ,
+		dragModifier = event.ctrlKey && !event.shiftKey ? "control" : !event.ctrlKey && event.shiftKey ? "shift" : event.ctrlKey && event.shiftKey ? "control-shift" : "default" ,
+		drag = [ "link", "plain", "transclude", "user" ] ;
+	if (dragSettings !== undefined) {
```
[the remark this answers is not in the document]
Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: consider it changed

**Comment by BurningTreeC** [comment]
@Jermolene 

I've removed the unnecessary regex tests because they're unnecessary
and made the control key be a ctrl key

Now the part with passing the plain titles on to droppable and dropzone widgets - I really don't have a clue how to do it differently. The downside with my method is the dragging between wikies that are not the same version: there will be an additional "Untitled" tiddler with an unnamed field containing the last tiddler title from the titles list

I hope it's clear why I want to pass the plain titles on - there would be no need for it if one uses the key that creates the plain titles for dragging into dropzones or droppable areas but this will cause mistakes

I cannot check the drag-target within draggable.js - that's why I came up with the current solution, I also had one that adds the plain string somewhere to $tw.dragInProgress but that doesn't allow handling inter-wiki-dragndrop

**Comment by Jermolene** [comment]
Thanks @BurningTreeC, and apologies for being slow on the uptake.

So, the problem is that the stringified list format used for the text representation of a drag is used both as a data format (e.g. when triggering drop actions) and as a presentation format (e.g. when dragging into a text field or into another app).

The solution here switches things so that the text representation of the drag is always treated as a presentation format, and introduces a new side channel for the underlying data format.

The trouble is that what we've ended up with seems brittle, and I worry that will make it harder to add further features to drag and drop. The backwards compatibility issue is a good signifier of a deep problem.

So, I think we need to take a step back and see if we can figure out a better solution before considering it for the core.

One approach that might work is to intercept drop events on our own text areas, grab the stringified list format, and then mutate it according to the key modifiers. There would be some additional work to get the iframe editor working.

**Comment by BurningTreeC** [comment]
@Jermolene 
Right, thanks for the new possible approach!

I could then put my code into wherever that place for intercepting drop events into text areas is and strip it down to only what's needed, maybe inside framed.js, is it the handleInputEvent?

That would be much better

**Comment by Jermolene** [comment]
Hi @BurningTreeC 

I can see a couple of possible approaches:

* A new global drop event handler in a startup module (the frame editor would need to be able to add the same event handler to the iframes that it creates)
* Updating the current drop handlers to detect text areas and apply the special handling

For the latter approach, there are a few places in widgets/droppable.js and widgets/dropzone.js where there is a conditional like this:

```
if(["TEXTAREA","INPUT"].indexOf(event.target.tagName) !== -1) {
```

**Comment by BurningTreeC** [comment]
I'd like to ask for something if we find a backwards compatible way:

We're having a discussion of dragndropping codeblocks in the google group [here](https://groups.google.com/forum/#!topic/TiddlyWiki/DaH8iadnPDg)

It was asked if one could also drag something into a textarea and have the text-content of the tiddler whose title gets dropped inserted in the textarea. I think it's doable here and would be an enhancement that makes sense.
If - I could implement a fifth identifier like "codeblock", that optionally can be assigned to a key (don't want to use more keys than shift and ctrl)

> Updating the current drop handlers to detect text areas and apply the special handling

I like using what's already there, implementing a new handler at startup fears me off a bit because I'm not experienced in doing it from scratch and I'm not able to invest the time to wrap my head around it

I feel like for the first approach I need help, with the second approach I could go on right away

**Comment by pmario** [comment]
>If - I could implement a fifth identifier like "codeblock", that optionally can be assigned to a key (don't want to use more keys than shift and ctrl)

Do you mean 

    ```
    code 
    ```

If yes, this will be a never ending story, since I want to add every possible wikitext modifier to my droped text, with the same key-combination.

**Comment by BurningTreeC** [comment]
@pmario 
No I don't mean that,
I mean that dragging the title holding the key that does "codeblock"
... collects the content of the tiddler with that title (it's text-field content) and inserts that to the textarea instead of the tiddler title being dragged

with that it'd be possible to create premade wikitext blocks and just drag them into the editor

**Comment by BurningTreeC** [comment]
@pmario sorry, naming it "codeblocks" is irritating. let's say "a block of wikitext stored somewhere"

**Comment by BurningTreeC** [comment]
@Jermolene to your two approaches:

Having looked at the droppable and dropzone widgets I've realized that it would require the textarea being a droppable zone or dropzone

The first approach now has gained my attention

I understand that the editor needs to pass the event handler on to the iframe so that inside framed.js it can be handled

I'll wrap my head around it

**Comment by Jermolene** [comment]
> Having looked at the droppable and dropzone widgets I've realized that it would require the textarea being a droppable zone or dropzone

That's not a problem; by default there is a dropzone in $:/PageTemplate that contains the entire page.

**Comment by Jermolene** [comment]
> It was asked if one could also drag something into a textarea and have the text-content of the tiddler whose title gets dropped inserted in the textarea. I think it's doable here and would be an enhancement that makes sense.

I think that can be done; the drop handler will have access to all of the data formats on the clipboard, including the full JSON representation, and so should be able to extract the text fields even when dragging between wikis.

However, I share @pmario's concern that we're in danger of overloading the modifier keys.

I also note that there is already a working way of dragging the text of a tiddler. Try this on tiddlywiki.com:

```
<$list filter="[tag[HelloThere]]">
<$draggable tiddler={{!!text}}><$text text=<<currentTiddler>>/></$draggable>
</$list>
```

**Comment by BurningTreeC** [comment]
@Jermolene 
> I also note that there is already a working way of dragging the text of a tiddler

right, didn't consider this - Thank you!
no need for what I asked for

**Comment by BurningTreeC** [comment]
>That's not a problem; by default there is a dropzone in $:/PageTemplate that contains the entire page.

Ok then the second approach looks best again, using the check for textarea or input in both droppable and dropzone

I'll give it a try now and report back then

**Comment by BurningTreeC** [comment]
@Jermolene 

I've seen that there are too many obstacles for me to make the needed changes by myself

Anyone who wants to work on this can do so, right?

I'll let my tries on this rest a bit until I have all pieces together to understand how to proceed

Simon

**Comment by Jermolene** [comment]
Hi @BurningTreeC I understand if you don't have time to pursue this, but please don't be discouraged by the unexpected difficulties we've encountered. Making big complicated changes is time consuming process that often has dead ends and backtracking before you find a solution. Features like the initial implementation of drag-and-drop or the editor toolbars are long journeys. This isn't in that league, but it's still fiddly.

Anyhow, I'm happy to help with the exploration process. Let me see if I can do a quick proof of concept...

**Comment by Jermolene** [comment]
Hi @BurningTreeC here's a very quick hack to demonstrate the technique. The following code replaces lines 113-115 of dropzone.js:

```
	// Check for being over a TEXTAREA or INPUT
	if(["TEXTAREA","INPUT"].indexOf(event.target.tagName) !== -1) {
		// Get the text being dragged
		var draggedText = event.dataTransfer.getData("text/plain"); // This works in Chrome, need to test other browsers
		// Transform the text being dragged
		draggedText = "{{" + draggedText + "}}";
		// Replace the selection with the dragged text
		var currentText = event.target.value;
		event.target.value = currentText.slice(0,event.target.selectionStart) + draggedText + currentText.slice(event.target.selectionStart);
		// Tell the browser that we handled the drop
		event.preventDefault();
		// Stop the drop ripple up to any parent handlers
		event.stopPropagation();
		return false;
	}
```

**Comment by BurningTreeC** [comment]
thanks @Jermolene 
I'm not that discouraged, I don't understand how to add the eventhandler and then how to pass it on to the iframe, and where or on what to listen for the drop event. I need to know how before proceeding with this and if I have to figure it out somehow, then it won't be these days or weeks

Thank you for the example, it works for input fields, as expected it does not work for framed textareas
I've seen that the framed engine listens for two events, and I thought that there the event handler must be passed on

**Comment by Jermolene** [comment]
No problem, @BurningTreeC this is a complex change and will need quite a lot of work to figure it out. I can give you some hints but I don't know the answer at this point.

The hack I posted above has a problem: it prevents you being able to choose the place where you want the dropped text to be inserted; by default, as you drag within a textarea/input you'll see the insertion point moving around to show where the text will go. Our drop event handler cancels the drop, and hence also cancels the movement of the insertion point, and the text is placed according to the previous position of the insertion point.

Those kind of obstacles are intrinsic to these sorts of changes, sadly. It's often two steps forward and one step back.

**Comment by BurningTreeC** [comment]
ok @Jermolene , I'll keep coming back to this whenever I think I'm on something that could be tried

**Comment by BurningTreeC** [comment]
Hi @Jermolene @pmario 
I was thinking about a different approach just for the droppable and dropzone widgets

Would it be acceptable to just report which key was held while dropping (this would work when keys get pressed after dragstart) by saving the key in a state tiddler, to be able to catch the key within a triggered action string like <$droppable actions="""<$list filter = statetiddler says shift then ... />(at the end delete the state tiddler)""" ?

That wouldn't enhance the editor-dragging, that's a downside to the previous thing - but as there's no solution in sight, this could be a (configurable) option and wouldn't limit the actions to just adding and appending prefixes and suffixes

Simon

**Comment by pmario** [comment]
I'm not 100% sure, what you exactly mean. ... but I'm in favor of the most flexible and simple version that works, in a way that there should be room for improvements in the future. ;)

**Comment by Jermolene** [comment]
Hi @BurningTreeC if you're suggesting exposing the ctrl/shift/alt key state to the actions of the droppable widget, then I think the best approach would be to pass them as variables, as we do the `actionTiddler` variable at the moment.

**Comment by BurningTreeC** [comment]
@Jermolene , that's what I mean in terms of functionality, passing them as variables is much better
@pmario that would be pretty simple and noone's forced to use them, what do you think?

**Comment by BurningTreeC** [comment]
@Jermolene @pmario 

I got a new approach working building on Jeremy's concept

At dragstart a class tc-window-dragging gets applied to document.body which allows to set iframe.tc-edit-texteditor { pointer-events: none; } ... that class gets removed at dragend.

That enables to listen for the drop on the text-editor iframe and change the title string accordingly to the modifier key held while dragging

Now this doesn't allow to move the text-cursor within the editor - the dropped title simply gets appended.

What I don't understand right now is how and where the texteditor itself listens for a drop.
How does the texteditor handle the drop of a dragged link, getting its title or text? I was looking for the code for this at different places but couldn't find it - it must be something trivial that I just don't see

Simon

**Comment by BurningTreeC** [comment]
Here's a demonstration http://dragndroptest.tiddlyspot.com/

normal dragging into a textarea behaves normal
if dragging with ctrl it adds {{  }} to the title
everything is done inside dropzone.js in handleDragOverEvent and handleDropEvent

**Comment by pmario** [comment]
@BurningTreeC .. As soon as I start dragging in FF I get the ROD "red screen of death" js error message

**Comment by BurningTreeC** [comment]
Ok i'll check what's messed up for ff

**Comment by BurningTreeC** [comment]
@pmario the deadly error is fixed, it works now on my latest firefox, too

**Comment by Jermolene** [comment]
> Now this doesn't allow to move the text-cursor within the editor - the dropped title simply gets appended.

I encountered that, too, and it's pretty annoying.

> What I don't understand right now is how and where the texteditor itself listens for a drop.
How does the texteditor handle the drop of a dragged link, getting its title or text? I was looking for the code for this at different places but couldn't find it - it must be something trivial that I just don't see

That behaviour is built in to the browser. The problem seems to be that the special behaviour of moving the insertion point before the drop isn't visible/accessible to JavaScript.

One point I meant to make before is that a lot of these difficulties stem from the fact that we're using the native browser drag and drop. It makes some things possible (like dragging from other browser windows) but makes some things much more complicated (like complex dragging within a single window). It's possible that some of the things you're exploring would be simpler if we evolved a framework for home-made dragging.

**Comment by BurningTreeC** [comment]
@Jermolene 

>  a framework for home-made dragging

That would be amazing -  I have looked at some frameworks, I've seen this one a while ago [draggable](https://github.com/Shopify/draggable), which seems to be a nice one

I've played a bit with my hack with the class on the document.body and I was able to move the cursor by toggling that class on/off/on (like crazy) inside one of the handler functions, I mean it was handledragleave. I admit that it's messy but it was fun. To work correctly this would need some tweaking of that class-on-off-toggling in all the handler-functions but I think it could be done .

But I really like the framework idea

Simon

**Comment by Jermolene** [comment]
Hi @BurningTreeC great, I'll be looking forward to seeing what you come up with :)

**Comment by BurningTreeC** [comment]
> I'll be looking forward to seeing what you come up with :)
@Jermolene it would be easier if the editor wouldn't be framed
Like replacing the framed engine with a basic codemirror, is that impossible?

> if we evolved a framework for home-made dragging
Were you talking about a self-made framework, something lightweight?
That would exceed my capabilities at the moment

I'll try my mad class-toggling for the framed editor and see where it leads
For the codemirror editor the solution would be the one from your hack, some comments above

all the best, Simon

**Comment by Jermolene** [comment]
> @Jermolene it would be easier if the editor wouldn't be framed. Like replacing the framed engine with a basic codemirror, is that impossible?

The iframe is there to enable the toolbar to work: with a naked textarea, clicking on a toolbar button would lose the focus and selection within the textarea.

The reason that I originally rejected incorporating CodeMirror was that it doesn't have great compatibility with the more exotic browsers (for many years it didn't work well on an iPad, for instance). It's also quite large.

> Were you talking about a self-made framework, something lightweight?
> That would exceed my capabilities at the moment

I was thinking of something homegrown, but it might be worth looking at off-the-shelf libraries, particularly to smooth out the differences between mouse and touch. An example use-case would be a slider that the user can drag to resize the sidebar relative to the story river: we need the ability to move a ghost of an element in 2D space, with the ability to constrain movement, and to update a specified tiddler when the drag completes.

> I'll try my mad class-toggling for the framed editor and see where it leads
> For the codemirror editor the solution would be the one from your hack, some comments above

Good luck.

**Comment by BurningTreeC** [comment]
> a slider that the user can drag to resize the sidebar relative to the story river
If I understand you right, I made something like that here:  http://panflex.tiddlyspot.com

There are many interesting frameworks out there, I'll try my luck and report back with news on the code

>Good luck.
Thanks

**Comment by Jermolene** [comment]
Hi @BurningTreeC 

> It's not framed like the native editor and indeed looses focus sometimes when clicking buttons in the toolbar, but it doesn't happen that often. It's a matter of timing as I experienced. Yes, it's annoying when it happens that focus on the selection gets lost.

Ouch! That's frustrating, but perhaps something that we can (eventually) diagnose and fix?

> I tried the codemirror panel-plugin where I put the toolbar in a codemirror top-panel and used the cm-native focus-method when a click is detected within that panel. It's definetily accessible from engine.js and maybe worth trying but my implementation was wrong and I had other priorities then

Your work has prompted me to think about experimenting with a new edit engine using a contenteditable DIV (helpfully, the selection is retained when clicking outside the contenteditable area). I did a lot of investigation of contenteditable back in 2010/11 when starting TW5, and found it was very powerful and straightforward to work with. But back then cross-browser compatibility was pretty poor; now, the implementations are a lot better.

> Thanks for your help on this, I learned a lot and I want to give it another try anytime soon but I think not before 5.1.16

It turned out to be a bit more ambitious than perhaps you'd anticipated! I'm very glad to see it completed.

**Pull request closed without merging**